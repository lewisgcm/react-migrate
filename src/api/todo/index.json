[
    {
        "userId": 1,
        "id": 1,
        "title": "clean the dishes",
        "completed": false
    },
    {
        "userId": 2,
        "id": 2,
        "title": "wash the floors",
        "completed": false
    },
    {
        "userId": 3,
        "id": 3,
        "title": "take out the rubbish",
        "completed": false
    }
]
