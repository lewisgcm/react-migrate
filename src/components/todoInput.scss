.TodoInput {
    input {
        width: 100%;
    }
}
