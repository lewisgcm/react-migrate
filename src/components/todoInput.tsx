import * as React from "react";

const style = require( "./todoInput.scss" );

export interface TodoInputProps {
    onAddTodo: (todo: string) => void;
}

interface TodoInputState {
    todo: string;
}

export class TodoInput extends React.Component<TodoInputProps, TodoInputState> {

    constructor( props: TodoInputProps ) {
        super( props );
        this.state = {
            todo: "",
        };
    }

    public render() {
        return <div className={style.TodoInput}>
            <input
                placeholder="Type a todo and press 'enter' to add"
                value={this.state.todo}
                onChange={( e ) => this.handleChange( e )}
                onKeyPress={( e ) => this.handleKeyPress( e )} >
            </input>
        </div>;
    }

    private handleKeyPress( e: React.KeyboardEvent ) {
        if ( e.key === "Enter" && this.state.todo.length > 0 ) {
            this.props.onAddTodo( this.state.todo );
            this.setState({ todo: "" });
        }
    }

    private handleChange( e: React.ChangeEvent<HTMLInputElement> ) {
        this.setState({
            todo: e.target.value,
        });
    }
}
