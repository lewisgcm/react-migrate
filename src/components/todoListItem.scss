.TodoListItem {
    width: 100%;
    line-height: 1.8em;

    span {
        border: 1px solid red;
        background-color: #FF8888;
        border-radius: 15px;
        height: 24px;
        width: 24px;
        text-align: center;
        font-size: .6em;
        float: right;
        cursor: pointer;
    }

    span:hover {
        background-color: #ffeeeeee;
    }
}
