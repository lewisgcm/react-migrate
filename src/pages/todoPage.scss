.TodoPage {
    h1 {
        color: red;
    }
}
