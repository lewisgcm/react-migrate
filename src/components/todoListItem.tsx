import * as React from "react";

import { Todo } from "../models/todo";

const style = require( "./todoListItem.scss" );

interface TodoListItemProps {
    todo: Todo;
    onDeleteTodo: (todo: Todo) => void;
}

interface TodoListItemState {
    todo: Todo;
}

export class TodoListItem extends React.Component<TodoListItemProps, TodoListItemState> {

    constructor( props: TodoListItemProps ) {
        super( props );
        this.state = {
            todo: props.todo,
        };
    }

    public render() {
        return <li className={style.TodoListItem}>
            {this.state.todo.title}
            <span onClick={( e ) => this.handleOnDelete()}>X</span>
        </li>;
    }

    private handleOnDelete() {
        this.props.onDeleteTodo( this.state.todo );
    }
}
