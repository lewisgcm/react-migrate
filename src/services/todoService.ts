import * as faker from "faker";
import { Observable, of } from "rxjs";
import { first } from "rxjs/operators";

import { Todo } from "../models/todo";
import { HttpClient } from "./httpClient";

export class TodoService {

    constructor(private httpClient: HttpClient) {
    }

    public getAll(): Observable<Todo[]> {
        return this.httpClient
            .Get<Todo[]>( "/api/todo/index.json" );
    }

    public get(id: number): Observable<Todo> {
        return this.httpClient
            .Get<Todo>( `/api/todo/${id}.json` );
    }

    // This is just a fake stub for adding a todo
    public add(title: string): Observable<Todo> {
        const fakeId = faker.random.number( 1000 );
        return of({
            id: fakeId,
            title,
            userId: fakeId,
            completed: false,
        }).pipe(
            first(),
        );
    }

    // Another fake stub for removing a todo
    public delete( todoId: number ): Observable<{}> {
        return of({}).pipe( first() );
    }
}
