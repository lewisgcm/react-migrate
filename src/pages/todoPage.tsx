import * as React from "react";
import * as ReactDOM from "react-dom";

import { TodoInput } from "../components/todoInput";
import { TodoListItem } from "../components/todoListItem";
import { Todo } from "../models/todo";
import { HttpClient } from "../services/httpClient";
import { TodoService } from "../services/todoService";

const style = require( "./todoPage.scss" );

export interface TodoPageProps {
    todoService: TodoService;
}

export interface TodoPageState {
    todos: Todo[];
}

export class TodoPage extends React.Component<TodoPageProps, TodoPageState> {

    constructor( props: TodoPageProps ) {
        super( props );
        this.state = {
            todos: [],
        };
    }

    public render() {
        return <div className={style.TodoPage}>
            <h1>My Todo Page</h1>
            <TodoInput onAddTodo={(t) => this.handleAddTodo(t)} ></TodoInput>
            <ul>
                {this.state.todos.map(
                    ( todo ) => <TodoListItem
                        key={todo.id}
                        todo={todo}
                        onDeleteTodo={(t) => this.handleDeleteTodo(t)} >
                    </TodoListItem>,
                )}
            </ul>
        </div>;
    }

    // Populate our todos when the page component is mounted
    public componentDidMount() {
        this.props.todoService
            .getAll()
            .subscribe(
                ( todos ) => {
                    this.setState({
                        todos,
                    });
                },
            );
    }

    // Handle the addition of a todo, i.e make an API call and save it.
    // We only update our state when we get a response.
    private handleAddTodo( todoTitle: string ) {
        this.props.todoService
            .add( todoTitle )
            .subscribe(
                ( todo ) => {
                    this.setState({
                        todos: [
                            ...this.state.todos,
                            todo,
                        ],
                    });
                },
            );
    }

    // Handle the deletion of a todo, i.e make the API call.
    // Again only update our state when its removed.
    private handleDeleteTodo( todo: Todo ) {
        this.props.todoService
            .delete( todo.id )
            .subscribe(
                () => {
                    this.setState({
                        todos: this.state
                            .todos
                            .filter( (t) => t.id !== todo.id ),
                    });
                },
            );
    }
}

export function createTodoPage( elementId: string ) {
    ReactDOM.render(
        <TodoPage todoService={new TodoService( new HttpClient() )}/>,
        document.getElementById( elementId ),
    );
}
