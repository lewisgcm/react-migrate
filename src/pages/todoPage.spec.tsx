import * as React from "react";
import { create } from "react-test-renderer";
import { of } from "rxjs";
import { instance, mock, when } from "ts-mockito";

import { TodoInput } from "../components/todoInput";
import { TodoListItem } from "../components/todoListItem";
import { Todo } from "../models/todo";
import { TodoService } from "../services/todoService";
import { TodoPage } from "./todoPage";

test( "Empty todo page", () => {
    const mockedTodoService = mock( TodoService );
    when( mockedTodoService.getAll() ).thenReturn( of([]) );

    const todoPage = create( <TodoPage todoService={instance( mockedTodoService )} /> );
    expect( todoPage ).toMatchSnapshot();
});

test( "Todo page with items", () => {
    const todo = {
        id: 0,
        userId: 1,
        title: "My todo",
        completed: true,
    } as Todo;

    const mockedTodoService = mock( TodoService );
    when( mockedTodoService.getAll() ).thenReturn( of([ todo ]) );

    const todoPage = create( <TodoPage todoService={instance( mockedTodoService )} /> );

    expect( todoPage ).toMatchSnapshot();
});

test( "Add todo", () => {
    const todo = {
        id: 0,
        userId: 1,
        title: "my todo",
        completed: true,
    } as Todo;
    const mockedTodoService = mock( TodoService );
    when( mockedTodoService.getAll() ).thenReturn( of([]) );
    when( mockedTodoService.add( "my todo" ) ).thenReturn( of( todo ) );

    const todoPage = create( <TodoPage todoService={instance( mockedTodoService )} /> );
    const input = todoPage.root
        .findByType( TodoInput )
        .find( (e) => e.type === "input" );

    input.props.onChange({ target: { value: "my todo" } });
    input.props.onKeyPress({ key: "Enter" });

    expect( todoPage ).toMatchSnapshot();
});

test( "Remove todo", () => {
    const todo = {
        id: 0,
        userId: 1,
        title: "my todo",
        completed: true,
    } as Todo;

    const mockedTodoService = mock( TodoService );
    when( mockedTodoService.getAll() ).thenReturn( of([ todo ]) );
    when( mockedTodoService.delete( 0 ) ).thenReturn( of({}) );

    const todoPage = create( <TodoPage todoService={instance( mockedTodoService )} /> );
    todoPage.root
        .findByType( TodoListItem )
        .find( (e) => e.type === "span" )
        .props.onClick({});

    expect( todoPage ).toMatchSnapshot();
});
