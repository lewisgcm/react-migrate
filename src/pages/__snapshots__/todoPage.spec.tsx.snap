// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add todo 1`] = `
<div
  className="TodoPage"
>
  <h1>
    My Todo Page
  </h1>
  <div
    className="TodoInput"
  >
    <input
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type a todo and press 'enter' to add"
      value=""
    />
  </div>
  <ul>
    <li
      className="TodoListItem"
    >
      my todo
      <span
        onClick={[Function]}
      >
        X
      </span>
    </li>
  </ul>
</div>
`;

exports[`Empty todo page 1`] = `
<div
  className="TodoPage"
>
  <h1>
    My Todo Page
  </h1>
  <div
    className="TodoInput"
  >
    <input
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type a todo and press 'enter' to add"
      value=""
    />
  </div>
  <ul />
</div>
`;

exports[`Remove todo 1`] = `
<div
  className="TodoPage"
>
  <h1>
    My Todo Page
  </h1>
  <div
    className="TodoInput"
  >
    <input
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type a todo and press 'enter' to add"
      value=""
    />
  </div>
  <ul />
</div>
`;

exports[`Todo page with items 1`] = `
<div
  className="TodoPage"
>
  <h1>
    My Todo Page
  </h1>
  <div
    className="TodoInput"
  >
    <input
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type a todo and press 'enter' to add"
      value=""
    />
  </div>
  <ul>
    <li
      className="TodoListItem"
    >
      My todo
      <span
        onClick={[Function]}
      >
        X
      </span>
    </li>
  </ul>
</div>
`;
